### PR TITLE
added rail-fence-cipher exercise

### PR DESCRIPTION
**Pull request by n8chz**

### CODE REVIEW

**Comment by wolf99** [comment]
Hi @n8chz ,

Welcome, and thanks for submitting an exercise!

Maybe this is still a work in progress?

If not, don't loose heart at the failing CI! There are only a few things left to do before this exercise is ready for review and merge.

1. We have updated recently to using a newer version (2.5.0) of the Unity test framework (mentioned [here](https://github.com/exercism/c/blob/master/docs/VERSIONS.md)). If you ensure your master branch of your fork is up to date with this repo you will be able to copy the correct version over from any of the other exercises.

2. Run `indent` against the repo. You can see that this is one of the items causing the CI to fail if you check the CI test results using the "details" link towards the bottom of this PR. `indent` can be run locally 🙌. There is information on how to run this in our Contributing documentation [here](https://github.com/exercism/c/blob/master/docs/CONTRIBUTING.md#coding-style)

3. The `make` tests are failing in the CI. I think this is foremost because the file naming is a little mixed up. Where you have the solution in `rail_fence_cipher.c` and the empty student file as `example.c`, this is the reverse of the expected structure for an exercise, just flip those file names around and the tests should look a lot healthier 😄 .

4. The `config.json` changes will also require an `"unlocked_by"` key so that Exercism knows where to place it in the syllabus, as a non-core exercise (you have set `"core"` to `false`) the exercise name used for the `"unlocked_by"` key, should refer to a core exercise (i.e. core exercises unlock zero or more non-core exercises, but non-core exercises do not unlock further exercises). Further information on the `config.json` is available in the Exercism documentation [here](https://github.com/exercism/docs/blob/master/language-tracks/configuration/exercises.md).

If you're stuck or not sure on anything add a comment here, we will be happy to help!

**Comment by wolf99** [comment]
Shoot you almost had it with commit 6378280 !
It was just you were trying to free the wrong thing because you had the arguments to `test_expected_value()` mixed up 😅

**Comment by wolf99** [comment]
Hi @n8chz 
Can you show the exact command you are running when you say you are running indent?
If I run on my machine it corrects everything back to the way it should be.
In your last commit there seems to have even been wholesale changes to the `config.json` which should not be needed as the existing version of this already passes tests.

I have seen issues with `indent` behaving differently than expected on my local machine, so it would be good to know which command you are running, with which `indent` (use `indent --version`), and in which environment (i.e. shell and OS)

Thanks!

**Comment by n8chz** [comment]
OK I just now ran `./indent.sh` followed by `bin/verify-indent` and both ran with no output to console, which I assume means OK. Followed that with a git commit which says `nothing to commit, working tree clean`, verify-indent run prior to previous commit (since which nothing has changed) had said
> Incorrect style detected:

> config.json

Changes to config.json are a result of running `bin/fetch-configlet` followed by `bin/configlet`, which I probably shouldn't have done.

bin/verify-indent contains a comment that says
> \# This script fails when run by Travis if the diff is not checked twice
> \# Unfortunately no better solution was found
After which it runs diff twice. So I can only assume that people who have trodden this ground before me have had Travis Trouble.

**Comment by n8chz** [comment]
Had to manually fix `if(` to `if (` in exercises/list-ops/src/example.c and then **not** run ./indent.sh or bin/verify-indent.

**Comment by ryanplusplus** [comment]
@wolf99 I think that the indent issue is the same one we saw a few months ago where `indent` on Mac and `indent` on newer Linux distros were not behaving the same.

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,5 @@ bin/configlet.exe
 build/
 **/*.out
 **/*.o
+.*.swp
```
Comment (wolf99) [inline .gitignore:6]:
```suggestion

```
Can we have this as a separate PR and keep this one related to this exercise only?

File: bin/verify-indent [hunk 2]
```diff
@@ -6,7 +6,7 @@ bash indent.sh
 # Unfortunately no better solution was found
 git diff --quiet
 
-if ! git diff --quiet
+if ! git diff
```
Comment (wolf99) [inline bin/verify-indent:9]: Was this removed for your debug purposes?
```suggestion
if ! git diff --quiet
```

File: config.json [hunk 3]
```diff
@@ -25,8 +25,8 @@
       "topics": [
         "control_flow_if_else_statements",
         "control_flow_loops",
-        "math",
-        "logic"
+        "logic",
+        "math"
```
Comment (wolf99) [inline config.json:29]:
```suggestion
        "math",
        "logic"
```
Not sure why these changes to unrelated exercises? If you'd like to tidy this up that's great but please do on a separate PR. This separation of PRs makes it easier to manage the repo's history

File: config.json [hunk 4]
```diff
@@ -38,8 +38,8 @@
       "difficulty": 1,
       "topics": [
         "control_flow_if_else_statements",
-        "math",
-        "logic"
+        "logic",
+        "math"
```
Comment (wolf99) [inline config.json:42]:
```suggestion
        "math",
        "logic"
```

File: config.json [hunk 5]
```diff
@@ -63,8 +63,8 @@
       "unlocked_by": "resistor-color",
       "difficulty": 1,
       "topics": [
-        "enums",
-        "arrays"
+        "arrays",
+        "enums"
```
Comment (wolf99) [inline config.json:67]:
```suggestion
        "enums",
        "arrays"
```

Comment (wolf99) [reply config.json:67]: same for all the rest of the changes in this file (apart from as related to this exercise). Not able to propose changes for all of them so it will require a local side change.

File: exercises/crypto-square/src/example.c [hunk 6]
```diff
@@ -18,8 +18,8 @@ static size_t mstrlen(const char *input)
 }
 
 /* get_cols_and_rows: calculate columns and rows for the input square */
-static void get_cols_and_rows_input(size_t len, size_t * cols_inp,
-                                    size_t * rows_inp)
+static void get_cols_and_rows_input(size_t len, size_t *cols_inp,
+                                    size_t *rows_inp)
```
Comment (wolf99) [inline exercises/crypto-square/src/example.c:22]:
```suggestion
static void get_cols_and_rows_input(size_t len, size_t * cols_inp,
                                    size_t * rows_inp)
```

File: exercises/linked-list/src/example.c [hunk 7]
```diff
@@ -49,12 +49,12 @@ struct list_item **new_list(void)
    return list;
 }
 
-bool is_list_empty(struct list_item ** list)
+bool is_list_empty(struct list_item **list)
```
Comment (wolf99) [inline exercises/linked-list/src/example.c:52]:
```suggestion
bool is_list_empty(struct list_item ** list)
```

File: exercises/linked-list/src/example.c [hunk 8]
```diff
@@ -49,12 +49,12 @@ struct list_item **new_list(void)
...
 {
    return (list == NULL || *list == NULL);
 }
 
-bool push(struct list_item ** list, ll_data_t item_data)
+bool push(struct list_item **list, ll_data_t item_data)
```
Comment (wolf99) [inline exercises/linked-list/src/example.c:57]:
```suggestion
bool push(struct list_item ** list, ll_data_t item_data)
```

File: exercises/linked-list/src/example.c [hunk 9]
```diff
@@ -75,7 +75,7 @@ bool push(struct list_item ** list, ll_data_t item_data)
    return result;
 }
 
-ll_data_t pop(struct list_item ** list)
+ll_data_t pop(struct list_item **list)
```
Comment (wolf99) [inline exercises/linked-list/src/example.c:78]:
```suggestion
ll_data_t pop(struct list_item ** list)
```

File: exercises/linked-list/src/example.c [hunk 10]
```diff
@@ -95,7 +95,7 @@ ll_data_t pop(struct list_item ** list)
    return result;
 }
 
-ll_data_t shift(struct list_item ** list)
+ll_data_t shift(struct list_item **list)
```
Comment (wolf99) [inline exercises/linked-list/src/example.c:98]:
```suggestion
ll_data_t shift(struct list_item ** list)
```

File: exercises/linked-list/src/example.c [hunk 11]
```diff
@@ -114,7 +114,7 @@ ll_data_t shift(struct list_item ** list)
    return result;
 }
 
-bool unshift(struct list_item ** list, ll_data_t item_data)
+bool unshift(struct list_item **list, ll_data_t item_data)
```
Comment (wolf99) [inline exercises/linked-list/src/example.c:117]:
```suggestion
bool unshift(struct list_item ** list, ll_data_t item_data)
```

File: exercises/list-ops/src/example.c [hunk 12]
```diff
@@ -35,7 +35,7 @@ list_t *append_list(list_t * list1, list_t * list2)
    return list;
 }
 
-list_t *filter_list(list_t * list, bool(*filter) (list_value_t value))
+list_t *filter_list(list_t * list, bool (*filter)(list_value_t value))
```
Comment (wolf99) [inline exercises/list-ops/src/example.c:38]:
```suggestion
list_t *filter_list(list_t * list, bool(*filter) (list_value_t value))
```

File: exercises/list-ops/src/example.c [hunk 13]
```diff
@@ -80,9 +80,9 @@ list_t *map_list(list_t * list, list_value_t(*map) (list_value_t value))
    return mapped;
 }
 
-list_value_t foldl_list(list_t * list, list_value_t initial,
-                        list_value_t(*foldl) (list_value_t value,
-                                              list_value_t initial))
+list_value_t
+foldl_list(list_t * list, list_value_t initial,
+           list_value_t(*foldl) (list_value_t value, list_value_t initial))
```
Comment (wolf99) [inline exercises/list-ops/src/example.c:85]:
```suggestion
list_value_t foldl_list(list_t * list, list_value_t initial,
                        list_value_t(*foldl) (list_value_t value,
                                                                      list_value_t initial))
```

File: exercises/list-ops/src/example.c [hunk 14]
```diff
@@ -93,9 +93,9 @@ list_value_t foldl_list(list_t * list, list_value_t initial,
    return initial;
 }
 
-list_value_t foldr_list(list_t * list, list_value_t initial,
-                        list_value_t(*foldr) (list_value_t value,
-                                              list_value_t initial))
+list_value_t
+foldr_list(list_t * list, list_value_t initial,
+           list_value_t(*foldr) (list_value_t value, list_value_t initial))
```
Comment (wolf99) [inline exercises/list-ops/src/example.c:98]:
```suggestion
list_value_t foldr_list(list_t * list, list_value_t initial,
                        list_value_t(*foldr) (list_value_t value,
                                              list_value_t initial))
```

File: exercises/list-ops/src/example.h [hunk 15]
```diff
@@ -18,7 +18,7 @@ list_t *new_list(size_t length, list_value_t values[]);
 list_t *append_list(list_t * list1, list_t * list2);
 
 // filter list returning only values that satisfy the filter function
-list_t *filter_list(list_t * list, bool(*filter) (list_value_t value));
+list_t *filter_list(list_t * list, bool (*filter)(list_value_t value));
```
Comment (wolf99) [inline exercises/list-ops/src/example.h:21]:
```suggestion
list_t *filter_list(list_t * list, bool(*filter) (list_value_t value));
```

File: exercises/list-ops/src/list_ops.h [hunk 16]
```diff
@@ -18,7 +18,7 @@ list_t *new_list(size_t length, list_value_t values[]);
 list_t *append_list(list_t * list1, list_t * list2);
 
 // filter list returning only values that satisfy the filter function
-list_t *filter_list(list_t * list, bool(*filter) (list_value_t value));
+list_t *filter_list(list_t * list, bool (*filter)(list_value_t value));
```
Comment (wolf99) [inline exercises/list-ops/src/list_ops.h:21]:
```suggestion
list_t *filter_list(list_t * list, bool(*filter) (list_value_t value));
```

File: exercises/pascals-triangle/src/example.c [hunk 17]
```diff
@@ -1,6 +1,6 @@
 #include "pascals_triangle.h"
 
-void free_triangle(size_t ** mem_block, size_t n)
+void free_triangle(size_t **mem_block, size_t n)
```
Comment (wolf99) [inline exercises/pascals-triangle/src/example.c:3]:
```suggestion
void free_triangle(size_t ** mem_block, size_t n)
```

File: exercises/pascals-triangle/test/test_pascals_triangle.c [hunk 18]
```diff
@@ -10,7 +10,7 @@ void tearDown(void)
 {
 }
 
-static bool check(size_t count, size_t expected[][count], size_t ** result)
+static bool check(size_t count, size_t expected[][count], size_t **result)
```
Comment (wolf99) [inline exercises/pascals-triangle/test/test_pascals_triangle.c:13]:
```suggestion
static bool check(size_t count, size_t expected[][count], size_t ** result)
```

File: exercises/pig-latin/src/example.c [hunk 19]
```diff
@@ -20,8 +20,8 @@ static bool starts_with_vowel_sound(const char *input)
        || (first_letter == 'y' && sec_letter == 't');
 }
 
-static void count_words_and_chars(const char *input, size_t * words,
-                                  size_t * chars)
+static void count_words_and_chars(const char *input, size_t *words,
+                                  size_t *chars)
```
Comment (wolf99) [inline exercises/pig-latin/src/example.c:24]:
```suggestion
static void count_words_and_chars(const char *input, size_t * words,
                                  size_t * chars)
```

OMG @ryanplusplus this spacing by indent on the pointers is cray cray. What is the logic on the first parameter not having a space after the asterisk but the rest of them doing so?

File: exercises/rail-fence-cipher/src/example.c [hunk 20]
```diff
@@ -0,0 +1,54 @@
+#include "rail_fence_cipher.h"
+#include <string.h>
+#include <stdlib.h>
+
+char *encode(char *text, size_t rails)
+{
+   size_t length = strlen(text);
+   char *ciphertext = (char *)calloc(length + 1, sizeof(char));
+   size_t cell_size = 2 * rails - 2;
+   if (!cell_size) {            // kludge
```
Comment (wolf99) [inline exercises/rail-fence-cipher/src/example.c:10]: Don't include kludges. The example should probably be best practice. If checking that `cell_size` is not equal to zero, then do that (it doesn't take many more characters 😉 ), as in 

```
if (cell_size != 0) {        // no kludges here :)
```

File: exercises/rail-fence-cipher/src/example.c [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+   }
+   return ciphertext;
+}
+
+char *decode(char *ciphertext, size_t rails)
+{
+   size_t length = strlen(ciphertext);
+   char *text = (char *)calloc(length + 1, sizeof(char));
+   size_t cell_size = 2 * rails - 2;
+   if (!cell_size) {            // kludge
```
Comment (wolf99) [inline exercises/rail-fence-cipher/src/example.c:33]: Same for this kludge

File: exercises/react/src/react.h [hunk 22]
```diff
@@ -4,8 +4,8 @@
 struct reactor;
 struct cell;
 
-typedef int (*compute1) (int);
-typedef int (*compute2) (int, int);
+typedef int (*compute1)(int);
+typedef int (*compute2)(int, int);
```
Comment (wolf99) [inline exercises/react/src/react.h:8]:
```suggestion
typedef int (*compute1) (int);
typedef int (*compute2) (int, int);
```

File: exercises/react/src/react.h [hunk 23]
```diff
@@ -19,7 +19,7 @@ struct cell *create_compute2_cell(struct reactor *, struct cell *,
 int get_cell_value(struct cell *);
 void set_cell_value(struct cell *, int new_value);
 
-typedef void (*callback) (void *, int);
+typedef void (*callback)(void *, int);
```
Comment (wolf99) [inline exercises/react/src/react.h:22]:
```suggestion
typedef void (*callback) (void *, int);
```

File: indent.sh [hunk 24]
```diff
@@ -19,7 +19,7 @@ fi
 
 # Run indent on every file
 # shellcheck disable=SC2046
-$INDENT -nbad -bap -nbc -bbo -hnl -br -brs -c33 -cd33 -ncdb -ce -ci4 -cli0 -d0 -di1 -nfc1 -i3 -nut -ip0 -l80 -lp -npcs -nprs -npsl -sai -saf -saw -ncs -nsc -sob -nfca -cp33 -ss -il1 $(find exercises/ -not -path '*/vendor/*' -name '*.c' -or -not -path '*/vendor/*' -name '*.h')
+$INDENT -v -nbad -bap -nbc -bbo -hnl -br -brs -c33 -cd33 -ncdb -ce -ci4 -cli0 -d0 -di1 -nfc1 -i3 -nut -ip0 -l80 -lp -npcs -nprs -npsl -sai -saf -saw -ncs -nsc -sob -nfca -cp33 -ss -il1 $(find exercises/ -not -path '*/vendor/*' -name '*.c' -or -not -path '*/vendor/*' -name '*.h')
```
Comment (wolf99) [inline indent.sh:22]: Again something for a separate PR.

```suggestion
$INDENT -nbad -bap -nbc -bbo -hnl -br -brs -c33 -cd33 -ncdb -ce -ci4 -cli0 -d0 -di1 -nfc1 -i3 -nut -ip0 -l80 -lp -npcs -nprs -npsl -sai -saf -saw -ncs -nsc -sob -nfca -cp33 -ss -il1 $(find exercises/ -not -path '*/vendor/*' -name '*.c' -or -not -path '*/vendor/*' -name '*.h')
```

**Comment by patricksjackson** [comment]
Well, I messed this one up. I rebased against master, updated with the CI version of indent and force pushed up to n8chz:master (as you'd need to do with a rebase). But I pushed my own master instead of the branch I was working on, so this was closed because the branch this was coming from was identical to exercism/c master.

The code changes are not lost, they're in a branch in my repo, I may just open a PR from there - https://github.com/patricksjackson/xc/tree/pr/446

The indent issue is because indent had a new release for the first time in 8 years (the ones here), and with Travis on the old version this will cause lots of problems.

**Comment by n8chz** [comment]
Ah, it's OK, you can have the issue, if it gets an exercise added to the track. Now that I know the one weird trick for getting past Travis, I plan to add many more. I want C to be one of the "big" tracks, like Python and Ruby.

**Comment by wolf99** [comment]
@n8chz You might be interested to see the exercism v3 that is currently being built (https://github.com/exercism/v3) which will apply across all tracks. Already there is some work on a new C track there.
